### PR TITLE
Privoxy in place of tinyproxy #1758

### PR DESCRIPTION
**Pull request by jay-to-the-dee**

### CODE REVIEW

File: Dockerfile
Comment (pkishino) [inline Dockerfile:54]: @haugene should we still keep tinyproxy as well?

File: Dockerfile
Comment (pkishino) [inline Dockerfile:40]: comment should be removed

File: Dockerfile
Comment (pkishino) [inline Dockerfile:39]: Hmm, I checked and it's a bit inconvenient that they don't have a "latest" tag or such to use for here..

File: Dockerfile
Comment (pkishino) [inline Dockerfile:99]: this should be optional depending on if it's enabled or not

File: tinyproxy/start.sh
Comment (pkishino) [inline tinyproxy/start.sh:51]: remove commented out code please

File: tinyproxy/start.sh
Comment (pkishino) [inline tinyproxy/start.sh:54]: configuration should still be possible, no?

**Review by pkishino — CHANGES_REQUESTED** [review]
A number of items still remaining..commented out code, references to tinyproxy @haugene  I would like your view on this, if we should drop this
also, documentation updates necessary etc
how to config etc..

**Comment by pkishino** [comment]
I also changed the base to dev, please rebase and make sure only your commits are included in the PR

File: Dockerfile
Comment (amosshapira) [reply Dockerfile:40]: I expect this comment is a blocker because the shell doesn't interpret the `\` at the end of the line as a continuation.

**Comment by amosshapira** [comment]
Can you add a description of this PR?

File: Dockerfile
Comment (amosshapira) [reply Dockerfile:39]: Yep. The closest I could find to automatically becoming aware of Privoxy releases is to parse their release feed from here https://www.privoxy.org/feeds/privoxy-releases.xml 

Here is a one-liner to extract the latest source URL from the RSS feed and give it to `curl` to download from:

```suggestion
    && curl -sL $(curl -s https://www.privoxy.org/feeds/privoxy-releases.xml | egrep -o -m 1 'https:\/\/www\.privoxy\.org\/sf-download-mirror\/Sources\/[0-9]+\.[0-9]+\.[0-9]+%20\(stable\)\/privoxy-[0-9]+\.[0-9]+\.[0-9]+-stable-src\.tar\.gz<' | sed -n -e 's/<$//' -e 1p) | tar -C /opt/privoxy --strip-components=2 -xz \
```

And BTW - 3.0.29 is already 3 releases behind the current latest as of today: https://www.privoxy.org/sf-download-mirror/Sources/3.0.32%20%28stable%29/

File: Dockerfile
Comment (amosshapira) [inline Dockerfile:46]: `make install-strip` will also strip symbols from the binary, making it a bit smaller.

Ref: https://www.privoxy.org/gitweb/?p=privoxy.git;a=blob;f=GNUmakefile.in;h=fcd44c4f449897df249686c034466a1da031e215;hb=HEAD#l841

**Review by amosshapira — CHANGES_REQUESTED** [review]
1. Suggestion to fetch the latest privoxy source from the RSS feed.
2. The comment in the middle of a `&&` chain breaks it.

**Comment by haugene** [comment]
> A number of items still remaining..commented out code, references to tinyproxy @haugene I would like your view on this, if we should drop this
> also, documentation updates necessary etc
> how to config etc..

I agree on the comments to code and structure. The PR also needs to be rebased as it's now affecting a lot of files that I don't think has anything to do with this. A documentation revamp would be great to go with this. We've had a proxy for a long time but not a good section on how/why we use it etc. Also curious for a description of this PR itself. I guess the suggestion is to change tinyproxy for privoxy because it is better, but I'm not familiar with any of the solutions and would like a summary or pitch somehow. Also if this would be relatively backwards compatible or not?

**Comment by ilike2burnthing** [comment]
Brief explanation here - https://github.com/haugene/docker-transmission-openvpn/issues/1758

**Comment by pkishino** [comment]
@jay-to-the-dee I hope you have not abandoned this and can help address the remaining issues and comments, hope to get this in as it def. looks like an improvement

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile:46]: 3.0.29 is intentional :) I found later releases crash after a few hours but 3.0.29 does not have this issue :)

**Comment by jay-to-the-dee** [comment]
Hi @pkishino not abandoned, just been very busy IRL this last week and wanted to let the discussion run it's course first :) I will try and remerge against `dev` branch later with the suggested changes (truthfully I'm more familiar with `stable` branch over `dev` at present)

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile:40]: Worked fine for me but I appreciate may not for others so removed :)

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile:54]: @pkishino Keeping TinyProxy or not will come down to how much the HTTP auth feature means to you/ other users (`WEBPROXY_USERNAME` and `WEBPROXY_PASSWORD`) This can be supported by setting Privoxy to go upstream to TinyProxy, in which case we would need both services. Otherwise Privoxy can run fine without TinyProxy, which is how I run my instance (and tunnels through the VPN just fine in the same way TinyProxy does).

This is because Privoxy does not contain built in support for HTTP proxy auth. (see https://www.privoxy.org/user-manual/config.html 7.4.9) and you would need to use the `enable-proxy-authentication-forwarding` feature of Privoxy as a part of sending the HTTP auth. provision upstream (to TinyProxy). Personally I'm fine without HTTP auth. which isn't that secure anyway and Privoxy instead supports acccess control by IP address via `permit-access` and `deny-access` directives, if you wanted to do that at the Privoxy level.

As for how to config, it is all done in text files so things can continue in a similair fashion to the present `sed`-ing methods, but I am not an expert at using `sed` so would appreciate some help there :)

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile:99]: Agreed, but I don't believe`Dockerfile` supports conditional EXPOSE keyword ? (Correct me if I'm wrong :) )

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile]: `install-strip` I have added, thanks for the great idea :) Once Privoxy becomes stable past 3.0.29 again, I will implement your XML parser 👍

File: tinyproxy/start.sh
Comment (jay-to-the-dee) [reply tinyproxy/start.sh]: Fixed in https://github.com/haugene/docker-transmission-openvpn/pull/1760/commits/0e3391120b0898d3588138a3608264d8eed3b503

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile:39]: 3.0.29 is my cherry picked release for now as I found later 'stable' versions crashing after a few hours of use. :) Would love to use your one liner in future though!

**Comment by amosshapira** [comment]
Sorry I feel like a dick for not mentioning these before:

Consider using Docker multi-stage to build the `privoxy` binary in a separate "build" image. That way you don't include the whole build chain in the final image. That's good both for size and security.

Ref: https://docs.docker.com/develop/develop-images/multistage-build/

File: Dockerfile
Comment (amosshapira) [reply Dockerfile:98]: `EXPOSE` is more of a documentation of the image rather than an actual command. The port is determined during run time.

Ref: https://docs.docker.com/engine/reference/builder/#expose

File: openvpn/tunnelUp.sh
Comment (amosshapira) [inline openvpn/tunnelUp.sh:43]: I know you just updated paths here, but I think this line will be much clearer with the reversed condition.
Also the test doesn't check if the file is executable or not:
```suggestion
[[ -f /opt/privoxy/start.sh && -x /opt/privoxy/start.sh ]] && /opt/privoxy/start.sh
```

File: privoxy/scripts/start.sh
Comment (amosshapira) [inline privoxy/scripts/start.sh:33]: Why is that? Inside the container, it might be running as the container's root.

File: privoxy/scripts/start.sh
Comment (amosshapira) [inline privoxy/scripts/start.sh:45]: weird indentation.

File: privoxy/scripts/start.sh
Comment (jay-to-the-dee) [reply privoxy/scripts/start.sh:33]: It's the exact same code that tinyproxy run under (i.e. not mine). But it's a pretty standard Unix thing, but can simpy remove the check if you want?

**Comment by jay-to-the-dee** [comment]
> Sorry I feel like a dick for not mentioning this before:
> 
> Consider using Docker multi-stage to build the `privoxy` binary in a separate "build" image. That way you don't include the whole build chain in the final image. That's good both for size and security.
> 
> Ref: https://docs.docker.com/develop/develop-images/multistage-build/
> 
> Edit: you are actually already doing this with `FloodUI` build, just do something similar for `privoxy`

That's a great idea, and will do :) It's something I've not done before (The FloodUI stuff is by the other docker-transmission-openvpn maintainer's), so will have a look into later 👍

File: privoxy/scripts/start.sh
Comment (jay-to-the-dee) [reply privoxy/scripts/start.sh:45]: I was trying to stick to the existing style and must have missed that, thanks!

**Comment by jay-to-the-dee** [comment]
> > Sorry I feel like a dick for not mentioning this before:
> > Consider using Docker multi-stage to build the `privoxy` binary in a separate "build" image. That way you don't include the whole build chain in the final image. That's good both for size and security.
> > Ref: https://docs.docker.com/develop/develop-images/multistage-build/
> > Edit: you are actually already doing this with `FloodUI` build, just do something similar for `privoxy`
> 
> That's a great idea, and will do :) It's something I've not done before (The FloodUI stuff is by the other docker-transmission-openvpn maintainer's), so will have a look into later 👍

@amosshapira check out https://github.com/haugene/docker-transmission-openvpn/pull/1760/commits/569c8b18eca5103df9bea4b5a42d37a650ea6a58 and see if you think I've done it right https://github.com/haugene/docker-transmission-openvpn/pull/1760/commits/1c37fbd1fc01fda2f3eb9f1379f9f97fba1e3071 :)

**Comment by pkishino** [comment]
still a bunch of @haugene commits in your commit log that you should remove ;)

File: openvpn/tunnelDown.sh
Comment (amosshapira) [inline openvpn/tunnelDown.sh:4]: Same comment here about reversing the condition to make it clearer:
```suggestion
[[ -f /opt/privoxy/stop.sh ]] && /opt/privoxy/stop.sh
```

File: privoxy/scripts/start.sh
Comment (amosshapira) [reply privoxy/scripts/start.sh:33]: never mind. Another time if there is a need.

**Comment by amosshapira** [comment]
> @amosshapira check out [569c8b1](https://github.com/haugene/docker-transmission-openvpn/commit/569c8b18eca5103df9bea4b5a42d37a650ea6a58) and [1c37fbd](https://github.com/haugene/docker-transmission-openvpn/commit/1c37fbd1fc01fda2f3eb9f1379f9f97fba1e3071) and see if you think I've done it right :)

The specific changes I asked for look good. Still have to update the base branch before merge is possible.

File: Dockerfile
Comment (jay-to-the-dee) [reply Dockerfile:104]: Will mark this as resolved then @pkishino :)

**Comment by jay-to-the-dee** [comment]
So the current PR design intent is to replace tinyproxy with privoxy as a drop in replacement, rather than trying to use tinyproxy as an upstream proxy. Are we okay to try it this way and see if anyone misses the (not very secure anyway) HTTP Proxy auth feature?

**Review by pkishino — APPROVED** [review]
@haugene I think this looks good, if you don't have any further comments, let's add this in

File: docs/index.md
Comment (pkishino) [inline docs/index.md:57]: Can you please also update the web-proxy.md with any necessary documentation changes? This can be done separately though

**Comment by amosshapira** [comment]
I don't have permission to add a preview but the code looks ok now.  Good job!

File: docs/index.md
Comment (jay-to-the-dee) [reply docs/index.md:57]: Have done with https://github.com/jay-to-the-dee/docker-transmission-openvpn/commit/f07e865be5d9e0d47fc40f6b3ba65ddb45b9c6b3 :)

**Comment by haugene** [comment]
I think this overall looks good and I'm looking forward to testing out Privoxy. This will be a good addition to the 4.0 release and good timing that we got it in on an already planned major version bump.

But something happened to the Dockerfile here that shouldn't have. We have changed the build setup for flood-for-transmission web ui and this pulled back the old and slow variant.

File: privoxy/scripts/start.sh
Comment (haugene) [inline privoxy/scripts/start.sh:16]: The reason why we had to check two places was that tinyproxy had different default location for config based on distro. And we used to have both ubuntu and alpine support so we had to figure out where it was. Should not be an issue anymore.

Comment (jay-to-the-dee) [reply privoxy/scripts/start.sh:16]: So alpine does have a privoxy build, but I chose to build manually because a) more up to date version than the repo b) Being able to compile in and try out the HTTPS MITM support which will greatly enhance Privoxy's effectiveness and ability to change requests on the now mostly HTTPS web (it's still experimental though, so have left it off for now, but a user could enable it simply by changing the config without recompiling)

**Pull request merged**
